File: app/src/main/java/com/shuneault/netrunnerdeckbuilder/fragments/ChoosePacksDialogFragment.java
package com.shuneault.netrunnerdeckbuilder.fragments;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import kotlin.Lazy;

import com.shuneault.netrunnerdeckbuilder.R;
import com.shuneault.netrunnerdeckbuilder.db.CardRepository;
import com.shuneault.netrunnerdeckbuilder.game.Format;
import com.shuneault.netrunnerdeckbuilder.game.Pack;

import java.util.ArrayList;

import static org.koin.java.standalone.KoinJavaComponent.inject;

public class ChoosePacksDialogFragment extends DialogFragment {


    private Format format;
    private ArrayList<String> mPackCodes = new ArrayList<>();
    private Lazy<CardRepository> repo = inject(CardRepository.class);

    public interface ChoosePacksDialogListener {
        void onChoosePacksDialogPositiveClick(DialogFragment dialog);
    }

    private ChoosePacksDialogListener mListener;

    private ArrayList<String> mPackNames = new ArrayList<>();
    private boolean[] arrChecks = new boolean[0];
    private ArrayList<String> selectedValues = new ArrayList<>();


    public ArrayList<String> getSelectedValues() {
        ArrayList<String> values = new ArrayList<>();
        for (int i = 0; i < arrChecks.length; i++) {
            if (arrChecks[i])
                values.add(mPackCodes.get(i));
        }
        return values;
    }

    public void setData(ArrayList<String> packFilter, Format format){
        this.selectedValues = packFilter;
        this.format = format;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        CardRepository cardRepo = repo.getValue();
        ArrayList<Pack> packs = cardRepo.getPacks(format);
        mPackCodes.clear();
        for (Pack p: packs) {
            mPackCodes.add(p.getCode());
            mPackNames.add(p.getName());
        }

        arrChecks = new boolean[mPackCodes.size()];
        for (int i = 0; i < mPackNames.size(); i++) {
            if (selectedValues.contains(mPackCodes.get(i))){
                arrChecks[i] = true;
            }
        }

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        // Set the dialog title
        builder.setTitle(R.string.choose_packs)
            // Specify the list array, the items to be selected by default (null for none),
            // and the listener through which to receive callbacks when items are selected
            .setMultiChoiceItems(mPackNames.toArray(new String[0]), arrChecks,
                    (dialog, which, isChecked) -> arrChecks[which] = isChecked)

            // Set the action buttons
            .setPositiveButton(R.string.ok, (dialog, id) -> mListener.onChoosePacksDialogPositiveClick(ChoosePacksDialogFragment.this))
            .setNeutralButton(R.string.reset, (dialog, id) -> {
                for (int i = 0; i < arrChecks.length; i++) {
                    arrChecks[i] = false;
                }
                mListener.onChoosePacksDialogPositiveClick(ChoosePacksDialogFragment.this);
            });

        return builder.create();
    }

    // Override the Fragment.onAttach() method to instantiate the NoticeDialogListener
    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        try {
            mListener = (ChoosePacksDialogListener) context;
        } catch (ClassCastException e) {
            throw new ClassCastException(getActivity().toString() + " must implement NoticeDialogListener");
        }
    }
}

File: app/src/main/java/com/shuneault/netrunnerdeckbuilder/fragments/ChoosePacksDialogFragment.kt
package com.shuneault.netrunnerdeckbuilder.fragments

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import androidx.fragment.app.DialogFragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.shuneault.netrunnerdeckbuilder.R
import com.shuneault.netrunnerdeckbuilder.db.CardRepository
import com.shuneault.netrunnerdeckbuilder.fragments.dialog.PackListRecyclerViewAdapter
import com.shuneault.netrunnerdeckbuilder.game.Format
import org.koin.android.ext.android.inject
import java.util.*

class ChoosePacksDialogFragment(val packFilter: ArrayList<String>, startFormat: Format, val allowFormatChange: Boolean) : DialogFragment() {
    interface ChoosePacksDialogListener {
        fun onChoosePacksDialogPositiveClick(dialog: DialogFragment?)
        fun onMyCollectionChosen(dialog: DialogFragment?)
    }

    var format: Format
    private lateinit var rv: RecyclerView
    private lateinit var adapter: PackListRecyclerViewAdapter
    private val repo: CardRepository by inject()

    private var mListener: ChoosePacksDialogListener? = null
    private val mPackCodes = ArrayList<String>()
    private val mPackNames = ArrayList<String>()
    private var selectedValues = ArrayList<String>()

    init {
        selectedValues = packFilter
        this.format = startFormat
    }

    fun getSelectedValues(): ArrayList<String> {
        return adapter.getSelectedValues().map { p -> p.code } as ArrayList<String>
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val view = layoutInflater.inflate(R.layout.fragment_item_list, null)
        rv = view.findViewById(R.id.pack_list)

        val spnFormat: Spinner = view.findViewById(R.id.spnFormat)
        spnFormat.adapter = ArrayAdapter<Format>(
                requireContext(),
                android.R.layout.simple_spinner_dropdown_item,
                repo.formats)
        spnFormat.setSelection(repo.formats.indexOf(format))
        spnFormat.isEnabled = allowFormatChange
        spnFormat.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View, position: Int, id: Long) {
                // update the deck format
                val newFormat = parent?.getItemAtPosition(position) as Format
                if (newFormat != format) {
                    format = newFormat
                    reloadPackList()
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }

        reloadPackList()

        val builder = MaterialAlertDialogBuilder(requireContext())
        builder.setTitle(R.string.choose_packs)
                .setView(view)
                // Set the action buttons
                .setPositiveButton(R.string.ok) { dialog: DialogInterface?, id: Int -> mListener!!.onChoosePacksDialogPositiveClick(this) }
                .setNegativeButton(R.string.reset) { dialog: DialogInterface?, id: Int ->
                    adapter.clearSelectedValues()
                    mListener!!.onChoosePacksDialogPositiveClick(this)
                }
                .setNeutralButton("Collection") { dialog: DialogInterface?, which: Int -> mListener!!.onMyCollectionChosen(this) }
        return builder.create()
    }

    private fun reloadPackList() {
        val packs = repo.getPacks(format)
        mPackCodes.clear()
        mPackNames.clear()
        for (p in packs) {
            mPackCodes.add(p.code)
            mPackNames.add(p.name)
        }

        // Set the pack list adapter
        rv.layoutManager = LinearLayoutManager(context)
        adapter = PackListRecyclerViewAdapter(packs, selectedValues)
        rv.adapter = adapter
    }

    // Override the Fragment.onAttach() method to instantiate the NoticeDialogListener
    override fun onAttach(context: Context) {
        super.onAttach(context)
        mListener = try {
            context as ChoosePacksDialogListener
        } catch (e: ClassCastException) {
            throw ClassCastException(activity.toString() + " must implement ChoosePacksDialogListener")
        }
    }
}